sdk/search/Azure.Search/src/Serialization/JsonExtensions.cs

```diff
@@ -0,0 +1,473 @@
+            SearchDocument document,
+            JsonSerializerOptions options)
+        {
+            Debug.Assert(writer != null);
+            Debug.Assert(document != null);
+            Debug.Assert(options != null);
+
+            writer.WriteStartObject();
+            foreach (string key in document.Keys)
+            {
+                writer.WritePropertyName(key);
+                object value = document[key];
+
+                // Write the value using JsonSerializer so all of our
+                // converters take effect
+                JsonSerializer.Serialize(
+                    writer,
+                    value,
+                    value?.GetType() ?? typeof(object),
+                    options);
+            }
+            writer.WriteEndObject();
+        }
+
+        /// <summary>
+        /// Verify that the next token to be read matches our expectation or
+        /// throw an exception otherwise.
+        /// </summary>
+        /// <param name="reader">The JSON reader.</param>
+        /// <param name="expected">The expected token type.</param>
+        public static void Expects(
+            this ref Utf8JsonReader reader,
+            JsonTokenType expected)
+        {
+            if (reader.TokenType != expected)
+            {
+                throw new JsonException(
+                    $"Expected {nameof(JsonTokenType)} " +
+                    expected.ToString() +
+                    ", not " +
+                    reader.TokenType.ToString());
+            }
+        }
+
+        /// <summary>
+        /// Verify and read the next property name.
+        /// </summary>
+        /// <param name="reader">The JSON reader.</param>
+        /// <returns>The name of the next property.</returns>
+        public static string ExpectsPropertyName(this ref Utf8JsonReader reader)
+        {
+            reader.Expects(JsonTokenType.PropertyName);
+            string name = reader.GetString();
+            reader.Read(); // Advance past property names after we read them
+            return name;
+        }
+
+        /// <summary>
+        /// Verify and read the next string value.
+        /// </summary>
+        /// <param name="reader">The JSON reader.</param>
+        /// <returns>The next string value.</returns>
+        public static string ExpectsString(this ref Utf8JsonReader reader)
+        {
+            reader.Expects(JsonTokenType.String);
+            return reader.GetString();
+        }
+
+        /// <summary>
+        /// Verify and read the next double? value.
+        /// </summary>
+        /// <param name="reader">The JSON reader.</param>
+        /// <returns>The next double? value.</returns>
+        public static double? ExpectsNullableDouble(
+            this ref Utf8JsonReader reader)
+        {
+            if (reader.TokenType == JsonTokenType.Null)
+            {
+                return null;
+            }
+            reader.Expects(JsonTokenType.Number);
```

Comment:
Is the purpose of this check to change the exception message and type of exception thrown? `GetDouble` already checks/throws.

Reply:
Yes - to change the exception message, but I'm going to remove these when we clean up the JSON handling  in the near future.